packages/vite/src/node/plugins/html.ts

```diff
@@ -600,7 +600,7 @@ export function buildHtmlPlugin(config: ResolvedConfig): Plugin {
           js = `import "${modulePreloadPolyfillId}";\n${js}`
         }
 
-        return js
+        return { code: js, moduleSideEffects: 'no-treeshake' }
```

Comment:
Would it be better that only turn off `treeshake` when it's mpa?

Reply:
I dont think the extra complexity would make this better, if it is actually possible to detect it reliably. It was always working as no-treeshake before, no? If this works, it should work both in mpa or spa.

Reply:
My point is that spa would lose the treeshake optimization after this PR, would this be expected?

Reply:
If that is the case, this isn't the correct fix. `moduleSideEffects` apply only to the JS module corresponding to the entry file though, and not to all other modules in your app. I pushed some tests to check that tree-shaking is still working [`232ca7c` (#13962)](https://github.com/vitejs/vite/pull/13962/commits/232ca7c7f1193e7cd2aaccb60877df6f5c83dbf4). Would you also try the cases you're worried about to see if my understanding of `moduleSideEffects` is right here?

Reply:
I have the same understanding with you. But one thing I couldn't understand is that shouldn't these [lines](https://github.com/vitejs/vite/pull/13962/files#diff-0fcc548f93ac98828516b065e7685e20845fd1fd087dad3074588a16ec8a22d3R7-R10) be reserved according to 

>`moduleSideEffects` apply only to the JS module corresponding to the entry file though

Reply:
That lines are in a module where `moduleSideEffects: 'no-treeshake'` is not used. This PR only sets it for the JS corresponding to each HTML entry point.

Reply:
Ah I see, I missed some contexts here. Then this pr lgtm